### DISCUSSION TITLE
How to save timezone information to a table with a column where timezone is enabled

### DISCUSSION THREAD

**Discussion by JohnnyRacer** [sqlalchemy/sqlalchemy#10078] [Usage Questions]
Hello, I've been trying to save a list of datetimes to a table with the schema defined below. The insertion works correct, but the times saved are always in UTC time and the timezone information seems to be discarded. Is there a way I can insert to the table and preserve the timezone information?
I am using SQLAlchemy version `2.0.18` and Postgres 15 with asyncpg.
```py
from sqlalchemy.ext.automap import automap_base
from sqlalchemy import ARRAY, JSON, Column, func, DateTime, insert
import datetime

Base = automap_base()

class TestTimeTable(Base):
    __tablename__ = 'testTimeTable'
    id = Column('id',Integer,autoincrement=True, primary_key=True)
    time_created = Column('createdAt' ,DateTime(timezone=True), server_default=func.now())
    time_updated = Column('updatedAt' ,DateTime(timezone=True), onupdate=func.now())
    times = Column('times', ARRAY(DateTime(timezone=True), nullable=False)

  # async_session is created via sessionmaker(create_async_engine(('CONNECTION_STRING'), echo=True,future=True))

# Some array of datetime objects with 

time_list = [datetime.datetime(2023, 7, 9, 10, 0, tzinfo=tzoffset(None, 25200)), datetime.datetime(2023, 7, 10, 10, 0, tzinfo=tzoffset(None, 25200)), datetime.datetime(2023, 7, 11, 10, 0, tzinfo=tzoffset(None, 25200)), datetime.datetime(2023, 7, 12, 10, 0, tzinfo=tzoffset(None, 25200)), datetime.datetime(2023, 7, 13, 10, 0, tzinfo=tzoffset(None, 25200)), datetime.datetime(2023, 7, 14, 10, 0, tzinfo=tzoffset(None, 25200)), datetime.datetime(2023, 7, 15, 10, 0, tzinfo=tzoffset(None, 25200)), datetime.datetime(2023, 7, 16, 10, 0, tzinfo=tzoffset(None, 25200)), datetime.datetime(2023, 7, 17, 10, 0, tzinfo=tzoffset(None, 25200)), datetime.datetime(2023, 7, 18, 10, 0, tzinfo=tzoffset(None, 25200)), datetime.datetime(2023, 7, 19, 10, 0, tzinfo=tzoffset(None, 25200))]

async with async_session() as session:  
    stmt =  insert(TestTimeTable, times=time_list)
    await session.execute(stmt)
    await session.commit()
# Entry is inserted correctly but the times column looks like the following where the timezone information is now lost.
# ["2023-07-09T17:00:00.000Z","2023-07-10T17:00:00.000Z","2023-07-11T17:00:00.000Z","2023-07-12T17:00:00.000Z","2023-07-13T17:00:00.000Z","2023-07-14T17:00:00.000Z","2023-07-15T17:00:00.000Z","2023-07-16T17:00:00.000Z","2023-07-17T17:00:00.000Z","2023-07-18T17:00:00.000Z","2023-07-19T17:00:00.000Z"]
```

**Comment by zzzeek** [comment]
I always need to double check these things so here is the full MCVE, the values round trip correctly.  it's just that "TIMESTAMP WITH TIMEZONE" doesn't actually store the timezone you passed, it just converts for the given timezone and stores in UTC.  see https://www.postgresql.org/docs/current/datatype-datetime.html

> For timestamp with time zone, the internally stored value is always in UTC (Universal Coordinated Time, traditionally known as Greenwich Mean Time, GMT). An input value that has an explicit time zone specified is converted to UTC using the appropriate offset for that time zone. If no time zone is stated in the input string, then it is assumed to be in the time zone indicated by the system's [TimeZone](https://www.postgresql.org/docs/current/runtime-config-client.html#GUC-TIMEZONE) parameter, and is converted to UTC using the offset for the timezone zone.

> When a timestamp with time zone value is output, it is always converted from UTC to the current timezone zone, and displayed as local time in that zone. To see the time in another time zone, either change timezone or use the AT TIME ZONE construct (see [Section 9.9.4](https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-ZONECONVERT)).


```py
import asyncio
import datetime

from dateutil.tz import tzoffset

from sqlalchemy import ARRAY
from sqlalchemy import Column
from sqlalchemy import DateTime
from sqlalchemy import func
from sqlalchemy import insert
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy.ext.asyncio import async_sessionmaker
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.orm import declarative_base

Base = declarative_base()


class TestTimeTable(Base):
    __tablename__ = "testTimeTable"
    id = Column("id", Integer, autoincrement=True, primary_key=True)
    time_created = Column(
        "createdAt", DateTime(timezone=True), server_default=func.now()
    )
    time_updated = Column(
        "updatedAt", DateTime(timezone=True), onupdate=func.now()
    )
    times = Column("times", ARRAY(DateTime(timezone=True)), nullable=False)


# Some array of datetime objects with

time_list = [
    datetime.datetime(2023, 7, 9, 10, 0, tzinfo=tzoffset(None, 25200)),
    datetime.datetime(2023, 7, 10, 10, 0, tzinfo=tzoffset(None, 25200)),
    datetime.datetime(2023, 7, 11, 10, 0, tzinfo=tzoffset(None, 25200)),
    datetime.datetime(2023, 7, 12, 10, 0, tzinfo=tzoffset(None, 25200)),
    datetime.datetime(2023, 7, 13, 10, 0, tzinfo=tzoffset(None, 25200)),
    datetime.datetime(2023, 7, 14, 10, 0, tzinfo=tzoffset(None, 25200)),
    datetime.datetime(2023, 7, 15, 10, 0, tzinfo=tzoffset(None, 25200)),
    datetime.datetime(2023, 7, 16, 10, 0, tzinfo=tzoffset(None, 25200)),
    datetime.datetime(2023, 7, 17, 10, 0, tzinfo=tzoffset(None, 25200)),
    datetime.datetime(2023, 7, 18, 10, 0, tzinfo=tzoffset(None, 25200)),
    datetime.datetime(2023, 7, 19, 10, 0, tzinfo=tzoffset(None, 25200)),
]

async def async_main():
    # engine is an instance of AsyncEngine
    engine = create_async_engine(
        "postgresql+asyncpg://scott:tiger@localhost/test",
        echo=True,
    )

    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.drop_all)
        await conn.run_sync(Base.metadata.create_all)

    async_session = async_sessionmaker(engine, expire_on_commit=False)

    async with async_session() as session:
        stmt = insert(TestTimeTable).values(times=time_list)
        await session.execute(stmt)
        await session.commit()


        times = await session.scalar(select(TestTimeTable.times))

        assert times == time_list

asyncio.run(async_main())

```

**Reply by zzzeek** [reply]
so to be clear, the answer for "how to save the timezones" would be to explicitly store it, either by:

1. storing the datetimes as a string (or I guess, other structure like JSON, though string is likely more straightforward) that includes the timezone , or 
2. storing the timezones in a separate column

**Reply by JohnnyRacer** [reply]
Thank you for the insight into how `DateTime` works. For an array of datetimes with differing timezones, I would have to create another array column to hold the timezone information so I can map it to the datetimes when they are accessed?

**Reply by zzzeek** [reply]
I think the funny part here is why you need to store those timezones with the timestamps ?    Usually timezones are associated with places (this event is at Dodgers stadium, so the timezone is PET) or users (I've set my preferred timezone to ET).

**Reply by JohnnyRacer** [reply]
I'm dealing with a table where the timezones of the datetimes in the list are not fixed, due to the source of the data. The closest analogy I can think of is for arranging delivery times across the country and putting those datetimes in a list, the timezones could be the same or differing based on the location of the delivery. Therefore I would need to store both the datetimes and timezones to ensure when the datetime is accessed, the correct local time is shown instead of UTC (The person accessing may or may not know the origin timezone of the original datetime, so I can't just store it as UTC and just convert it directly since their timezone might be mismatched) .

**Reply by zzzeek** [reply]
but these dates line up with places then, right?  I mean sure.   store a separate array with the timezones, that's what you'd have to do if there is no other way to represent the total data you have.

**Reply by CaselIT** [reply]
> Thank you for the insight into how `DateTime` works. For an array of datetimes with differing timezones, I would have to create another array column to hold the timezone information so I can map it to the datetimes when they are accessed?

You can opt to use json instead of an array of datetimes. You can then format the datetimes as iso strings, that do include the timezone.
That's a bit more cumbersome to use in the db and you would need to manually parse the dates back to datetime in the application though
